payjoin/src/hpke.rs

```diff
@@ -0,0 +1,283 @@
+    fn from(value: HpkeKeyPair) -> Self { (value.0, value.1) }
+}
+
+impl HpkeKeyPair {
+    pub fn gen_keypair() -> Self {
+        let (sk, pk) = <SecpK256HkdfSha256 as hpke::Kem>::gen_keypair(&mut OsRng);
+        Self(HpkeSecretKey(sk), HpkePublicKey(pk))
+    }
+    pub fn secret_key(&self) -> &HpkeSecretKey { &self.0 }
+    pub fn public_key(&self) -> &HpkePublicKey { &self.1 }
+}
+
+#[derive(Clone, PartialEq, Eq)]
+pub struct HpkeSecretKey(pub SecretKey);
+
+impl Deref for HpkeSecretKey {
+    type Target = SecretKey;
+
+    fn deref(&self) -> &Self::Target { &self.0 }
+}
+
+impl core::fmt::Debug for HpkeSecretKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "SecpHpkeSecretKey({:?})", self.0.to_bytes())
+    }
+}
+
+impl serde::Serialize for HpkeSecretKey {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_bytes(&self.0.to_bytes())
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for HpkeSecretKey {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let bytes = Vec::<u8>::deserialize(deserializer)?;
+        Ok(HpkeSecretKey(
+            SecretKey::from_bytes(&bytes)
+                .map_err(|_| serde::de::Error::custom("Invalid secret key"))?,
+        ))
+    }
+}
+
+#[derive(Clone, PartialEq, Eq)]
+pub struct HpkePublicKey(pub PublicKey);
+
+impl HpkePublicKey {
+    pub fn to_compressed_bytes(&self) -> [u8; 33] {
+        let compressed_key = bitcoin::secp256k1::PublicKey::from_slice(&self.0.to_bytes())
+            .expect("Invalid public key from known valid bytes");
+        compressed_key.serialize()
+    }
+
+    pub fn from_compressed_bytes(bytes: &[u8]) -> Result<Self, HpkeError> {
+        let compressed_key = bitcoin::secp256k1::PublicKey::from_slice(bytes)?;
+        Ok(HpkePublicKey(PublicKey::from_bytes(
+            compressed_key.serialize_uncompressed().as_slice(),
+        )?))
+    }
+}
+
+impl Deref for HpkePublicKey {
+    type Target = PublicKey;
+
+    fn deref(&self) -> &Self::Target { &self.0 }
+}
+
+impl core::fmt::Debug for HpkePublicKey {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "SecpHpkePublicKey({:?})", self.0)
+    }
+}
+
+impl serde::Serialize for HpkePublicKey {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        serializer.serialize_bytes(&self.0.to_bytes())
+    }
+}
+
+impl<'de> serde::Deserialize<'de> for HpkePublicKey {
+    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+    where
+        D: serde::Deserializer<'de>,
+    {
+        let bytes = Vec::<u8>::deserialize(deserializer)?;
+        Ok(HpkePublicKey(
+            PublicKey::from_bytes(&bytes)
+                .map_err(|_| serde::de::Error::custom("Invalid public key"))?,
```

Comment:
If this is meant to be a compressed pk I would specify that in the err. 
Ignore if thats not the case

```suggestion
                .map_err(|_| serde::de::Error::custom("Invalid compressed public key"))?,
```

Reply:
It's meant to be an *uncompressed* pk